### PR TITLE
Feature/update tree cover gain

### PR DESCRIPTION
**Pull request by pjosh**
## Overview

The Tree cover gain widget that is currently just a dynamic sentence should be extended. A ranked list should be added to the whitespace in that widget showing 5 items at a time.

- As with all lists, the list items should be a link to their relevant country page.

- At the Admin-0 level, the ranked list shows the selected country in relation to its neighbouring ranked countries.

- At the admin-1 level, it shows the current selected admin-1 in relation to its neighbouring ranked admin-1 regions.

- At the admin-2 level, it shows the currently selected admin-2 in relation to its neighbouring ranked areas.

- A selector to change units between Ha and % should be added to the settings. This will affect whether or not we show rank order based on % or ha.

## Demo

kapture 2018-01-17 at 12 18 41

### CODE REVIEW

File: app/javascript/pages/country/widget/widgets/widget-tree-gain/widget-tree-gain-actions.js [hunk 1]
```diff
@@ -14,15 +14,24 @@ const getTreeGain = createThunkAction(
...
       axios
-        .all([getGain({ ...params }), getExtent({ ...params })])
+        .all([
+          getGain({ ...params }),
+          getExtent({ ...params }),
+          getGainRanking({ ...params })
+        ])
         .then(
-          axios.spread((gainResponse, extentResponse) => {
+          axios.spread((gainResponse, extentResponse, gainRankingResponse) => {
```
Comment (edbrett) [inline app/javascript/pages/country/widget/widgets/widget-tree-gain/widget-tree-gain-actions.js]: Im not sure an extra query is needed here. As we have to create a ranking list, surely a single fetch for all countries data is needed and then the selectors do the work?

File: app/javascript/pages/country/widget/widgets/widget-tree-gain/widget-tree-gain-actions.js [hunk 2]
```diff
@@ -35,6 +44,28 @@ const getTreeGain = createThunkAction(
   }
 );
 
+const getProcessedRankingData = (rawData, params) => {
```
Comment (edbrett) [inline app/javascript/pages/country/widget/widgets/widget-tree-gain/widget-tree-gain-actions.js]: This looks like a selector function to me. Lets keep data mapping to the selectors.

File: app/javascript/pages/country/widget/widgets/widget-tree-gain/widget-tree-gain-component.js [hunk 3]
```diff
@@ -2,24 +2,51 @@ import React, { PureComponent } from 'react';
...
+import WidgetNumberedList from 'pages/country/widget/components/widget-numbered-list';
+import NoContent from 'components/no-content';
+import COLORS from 'pages/country/data/colors.json';
 
 import './widget-tree-gain-styles.scss';
 
 class WidgetTreeCoverGain extends PureComponent {
   render() {
-    const { getSentence } = this.props;
+    const { locationNames, loading, data, settings, getSentence } = this.props;
```
Comment (edbrett) [inline app/javascript/pages/country/widget/widgets/widget-tree-gain/widget-tree-gain-component.js]: We no longer need locationNames and loading. These are held in the parent 'Widget' component.

File: app/javascript/pages/country/widget/widgets/widget-tree-gain/widget-tree-gain-selectors.js [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+export const getSortedData = createSelector(
+  [getData, getUnit, getLocation, getLocationsMeta],
+  (data, unit, location, meta) => {
+    if (!data || isEmpty(data) || !meta || isEmpty(meta)) return null;
+    const dataMapped = [];
+    data.forEach(d => {
+      const region = meta.find(l => d.id === l.value);
+      let path = '/country/';
+      if (location.subRegion) {
```
Comment (edbrett) [inline app/javascript/pages/country/widget/widgets/widget-tree-gain/widget-tree-gain-selectors.js]: We are starting to do this a few times. Can we add a util for this maybe?

File: app/javascript/pages/country/widget/widgets/widget-tree-gain/widget-tree-gain.js [hunk 5]
```diff
@@ -8,13 +8,35 @@ import { getActiveFilter } from 'pages/country/widget/widget-selectors';
...
-  loading: widgetTreeGain.loading || ownProps.isMetaLoading,
-  gain: widgetTreeGain.data.gain,
-  extent: widgetTreeGain.data.extent
-});
+const mapStateToProps = ({ location, widgetTreeGain, countryData }) => {
+  const { isCountriesLoading, isRegionsLoading } = countryData;
+  let meta = [];
+  if (location.payload.subRegion) {
+    meta = countryData.subRegions;
+  } else if (location.payload.region) {
```
Comment (edbrett) [inline app/javascript/pages/country/widget/widgets/widget-tree-gain/widget-tree-gain.js]: You can avoid an elseif case by setting meta = countryData.countries first

File: app/javascript/pages/country/widget/widgets/widget-tree-gain/widget-tree-gain.js [hunk 6]
```diff
@@ -8,13 +8,35 @@ import { getActiveFilter } from 'pages/country/widget/widget-selectors';
...
+    meta = countryData.regions;
+  } else {
+    meta = countryData.countries;
+  }
+
+  return {
+    loading: widgetTreeGain.loading || isCountriesLoading || isRegionsLoading,
+    data: {
+      gain: widgetTreeGain.data.gain,
+      extent: widgetTreeGain.data.extent,
```
Comment (edbrett) [inline app/javascript/pages/country/widget/widgets/widget-tree-gain/widget-tree-gain.js]: I have been playing with putting the dynamic sentence calculation inside the selectors. Seeing as we a re creating complicated structures for getting the data in the correct format, maybe it is easier this. Then we don't have to keep passing the data around, and we can cache the data to reduce recalculation. Thoughts?

File: app/javascript/pages/country/widget/widgets/widget-tree-gain/widget-tree-gain.js [hunk 7]
```diff
@@ -40,7 +63,7 @@ class WidgetTreeGainContainer extends PureComponent {
   }
 
   getSentence = () => {
```
Comment (edbrett) [inline app/javascript/pages/country/widget/widgets/widget-tree-gain/widget-tree-gain.js]: See comment above about putting in the selectors. I think it is better as then we only pass the data from mapStateToProps once and dont have to reassign in the container.

File: app/javascript/services/forest-data.js [hunk 8]
```diff
@@ -98,3 +107,28 @@ export const getFAOExtent = ({ country, period }) => {
     .replace('{period}', period);
   return axios.get(url);
 };
+
+export const getGainRanking = ({
```
Comment (edbrett) [inline app/javascript/services/forest-data.js]: Again, unsure if this is needed. I have done something similar in for the FAO reforestation widget which now only uses one query and uses just selectors. Take a look and let me know your thoughts.

**Comment by benlaken** [comment]
Functionally it works as it should now 👍

File: app/javascript/pages/country/data/widgets-config.json [hunk 9]
```diff
@@ -72,12 +72,16 @@
...
+      "selectors": ["indicators", "thresholds", "extentYears", "units"],
       "type": "gain"
     },
     "settings": {
       "indicator": "gadm28",
-      "threshold": "0"
+      "threshold": "0",
+      "unit": "ha",
+      "extentYear": 2010,
+      "pageSize": 5,
```
Comment (edbrett) [inline app/javascript/pages/country/data/widgets-config.json]: I guess as we aren't using pages we dont need these. I made an update to the numbered list comp in my branch to handle this better but obviously you dont have this here. Maybe not a concern for now.

File: app/javascript/services/forest-data.js [hunk 10]
```diff
@@ -27,6 +29,13 @@ const getLocationQuery = (country, region, subRegion) =>
     subRegion ? ` AND adm2 = ${subRegion}` : ''
   }`;
 
+const getRankingLocationQuery = (country, region, subRegion) =>
```
Comment (edbrett) [inline app/javascript/services/forest-data.js]: I think we had a helper for this (getLocationQuery), is there a difference?

Comment (pjosh) [reply app/javascript/services/forest-data.js]: oh! I didn't know it. Of course, I'm going to unify it.

File: app/javascript/pages/country/data/widgets-config.json (hunk 9 above)
Comment (pjosh) [reply app/javascript/pages/country/data/widgets-config.json]: nice! let me remove it

**Pull request merged**